ext/bigdecimal/bigdecimal.c

```diff
@@ -2199,33 +2148,39 @@ BigDecimal_div2(VALUE self, VALUE b, VALUE n)
 
     /* div in BigDecimal sense */
     ix = check_int_precision(n);
-    if (ix == 0) {
-        return BigDecimal_div(self, b);
-    }
-    else {
-        Real *res = NULL;
-        Real *av = NULL, *bv = NULL, *cv = NULL;
-        size_t mx = ix + VpBaseFig()*2;
-        size_t b_prec = ix;
-        size_t pl = VpSetPrecLimit(0);
-
-        GUARD_OBJ(cv, NewZeroWrapLimited(1, mx + VpBaseFig()));
-        GUARD_OBJ(av, GetVpValue(self, 1));
+
+    pl = VpSetPrecLimit(0);
+
+    GUARD_OBJ(av, GetVpValue(self, 1));
+    if (RB_FLOAT_TYPE_P(b) && ix > BIGDECIMAL_DOUBLE_FIGURES) {
         /* TODO: I want to refactor this precision control for a float value later
          *       by introducing an implicit conversion function instead of
          *       GetVpValueWithPrec.  */
-        if (RB_FLOAT_TYPE_P(b) && b_prec > BIGDECIMAL_DOUBLE_FIGURES) {
-            b_prec = BIGDECIMAL_DOUBLE_FIGURES;
-        }
-        GUARD_OBJ(bv, GetVpValueWithPrec(b, b_prec, 1));
-        mx = av->Prec + bv->Prec + 2;
-        if (mx <= cv->MaxPrec) mx = cv->MaxPrec + 1;
-        GUARD_OBJ(res, NewZeroWrapNolimit(1, (mx * 2  + 2)*VpBaseFig()));
-        VpDivd(cv, res, av, bv);
-        VpSetPrecLimit(pl);
-        VpLeftRound(cv, VpGetRoundMode(), ix);
-        return VpCheckGetValue(cv);
+        GUARD_OBJ(bv, GetVpValueWithPrec(b, BIGDECIMAL_DOUBLE_FIGURES, 1));
     }
+    else {
+        GUARD_OBJ(bv, GetVpValueWithPrec(b, ix, 1));
+    }
+
+    if (ix == 0) {
+        ssize_t a_prec, b_prec;
+        BigDecimal_count_precision_and_scale(av->obj, &a_prec, NULL);
+        BigDecimal_count_precision_and_scale(bv->obj, &b_prec, NULL);
+        ix = ((a_prec > b_prec) ? a_prec : b_prec) + BIGDECIMAL_DOUBLE_FIGURES;
+        if (2 * BIGDECIMAL_DOUBLE_FIGURES > ix)
+            ix = 2 * BIGDECIMAL_DOUBLE_FIGURES;
+    }
+
+    // VpDivd needs 2 extra DECDIGs. One more is needed for rounding.
+    GUARD_OBJ(cv, NewZeroWrapLimited(1, ix + 3 * VpBaseFig()));
+
+    mx = bv->Prec + cv->MaxPrec - 1;
+    if (mx <= av->Prec) mx = av->Prec + 1;
```

Comment:
Correct allocation size calculation of res.
VpDivd requires `word_a < word_r && word_b + word_c - 2 < word_r`